File: connectors/connector-trello/src/main/scala/com/taskadapter/connector/trello/GTaskToTrello.scala
package com.taskadapter.connector.trello

import java.util.Date

import com.julienvey.trello.domain.{Card, TList}
import com.taskadapter.connector.common.data.ConnectorConverter
import com.taskadapter.connector.definition.exception.FieldConversionException
import com.taskadapter.connector.definition.exceptions.ConnectorException
import com.taskadapter.model.{Children, Description, DueDate, GTask, Id, Key, ParentKey, Relations, SourceSystemId, Summary}
import org.slf4j.LoggerFactory

import scala.collection.JavaConverters._

class GTaskToTrello(config:TrelloConfig, listCache: ListCache) extends ConnectorConverter[GTask, Card] {
  val logger = LoggerFactory.getLogger(classOf[GTaskToTrello])

  override def convert(source: GTask): Card = {
    val card = new Card
    card.setId(source.getKey)
    card.setIdBoard(config.getBoardId)
    source.getFields.asScala.foreach { e =>
      val field = e._1
      val value = e._2
      try {
        field match {
          case Children => // processed in another place
          case Id => // ignore ID field because it does not need to be provided when saving
          case Key => // processed in [[DefaultValueSetter]]
          case SourceSystemId => // processed in [[DefaultValueSetter]]
          case ParentKey => // processed above
          case Relations => // processed in another place

          case TrelloField.listId =>
            card.setIdList(value.asInstanceOf[String])
          case TrelloField.listName =>
            val listName = value.asInstanceOf[String]
            val listId = listCache.getListIdByName(listName)
            if (listId.isDefined) {
              card.setIdList(listId.get)
            } else {
              throw new ConnectorException(
                s"Trello list with name '$listName' is not found on the requested Trello Board (board ID ${config.getBoardId} )")
            }
          case Summary => card.setName(value.asInstanceOf[String])
          case Description => card.setDesc(value.asInstanceOf[String])
          case DueDate => card.setDue(value.asInstanceOf[Date])
          case _ => logger.warn(s"Unknown field in GTask: $field. Skipping it")
        }
      } catch {
        case e: ConnectorException => throw e
        case e: Exception => throw new FieldConversionException(TrelloConnector.ID, field, value, e.getMessage)
      }
    }
    card
  }
}

File: connectors/connector-trello/src/main/java/com/taskadapter/connector/trello/GTaskToTrello.java
package com.taskadapter.connector.trello;

import com.julienvey.trello.domain.Card;
import com.taskadapter.connector.common.data.ConnectorConverter;
import com.taskadapter.connector.definition.exception.FieldConversionException;
import com.taskadapter.connector.definition.exceptions.ConnectorException;
import com.taskadapter.model.Children;
import com.taskadapter.model.Description;
import com.taskadapter.model.DueDate;
import com.taskadapter.model.Field;
import com.taskadapter.model.GTask;
import com.taskadapter.model.Id;
import com.taskadapter.model.Key;
import com.taskadapter.model.ParentKey;
import com.taskadapter.model.Relations;
import com.taskadapter.model.SourceSystemId;
import com.taskadapter.model.Summary;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Date;
import java.util.Map;

public class GTaskToTrello implements ConnectorConverter<GTask, Card> {
    private static final Logger logger = LoggerFactory.getLogger(GTaskToTrello.class);

    private final TrelloConfig config;
    private final ListCache listCache;

    public GTaskToTrello(TrelloConfig config, ListCache listCache) {
        this.config = config;
        this.listCache = listCache;
    }

    public Card convert(GTask source) throws ConnectorException {
        var card = new Card();
        card.setId(source.getKey());
        card.setIdBoard(config.getBoardId());
        for (Map.Entry<Field<?>, Object> entry : source.getFields().entrySet()) {
            var field = entry.getKey();
            try {
                processField(card, field, entry.getValue());
            } catch (ConnectorException e) {
                throw e;
            } catch (Exception e) {
                throw new FieldConversionException(TrelloConnector.ID, entry.getKey(), entry.getValue(), e.getMessage());
            }
        }
        return card;
    }

    private void processField(Card card, Field<?> field, Object value) throws ConnectorException {
        if (field instanceof Children) {
            // processed in another place
            return;
        }
        if (field instanceof Id) { // ignore ID field because it does not need to be provided when saving
            return;
        }
        if (field instanceof Key) { // processed in <<DefaultValueSetter>>
            return;
        }
        if (field instanceof SourceSystemId) { // processed in <<DefaultValueSetter>>
            return;
        }
        if (field instanceof ParentKey) { // processed above
            return;
        }
        if (field instanceof Relations) { // processed in another place
            return;
        }

        if (field.equals(TrelloField.listId)) {
            card.setIdList((String) value);
            return;
        }
        if (field.equals(TrelloField.listName)) {
            var listName = (String) value;
            var listId = listCache.getListIdByName(listName);
            if (listId.isPresent()) {
                card.setIdList(listId.get());
            } else {
                throw new ConnectorException(
                        "Trello list with name '" + listName + "' is not found on the requested Trello Board (board ID " + config.getBoardId() + ")");
            }
        }
        if (field instanceof Summary) {
            card.setName((String) value);
            return;
        }
        if (field instanceof Description) {
            card.setDesc((String) value);
            return;
        }
        if (field instanceof DueDate) {
            card.setDue((Date) value);
            return;
        }
        logger.warn("Unknown field in GTask: " + field + ". Skipping it");
    }
}
